From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Myrrh Periwinkle <myrrhperiwinkle@qtmlabs.xyz>
Date: Tue, 17 Jun 2025 10:42:46 +0700
Subject: Implement EGL_ANGLE_platform_angle_device_id on Linux

This allows the Chromium Ozone/Wayland backend to specify the specific
device that should be used for the EGL surfaceless display. This is
useful for ensuring that ANGLE use the same underlying GPU as the GBM
device for OzoneImageBacking to avoid accidental cross-device resource
sharing.

diff --git a/src/common/vulkan/vulkan_icd.cpp b/src/common/vulkan/vulkan_icd.cpp
index 4daed7da72bc7d256710ffa788b8e0c509be5483..39b3a45545244c555243b9af440d6babc56e1730 100644
--- a/src/common/vulkan/vulkan_icd.cpp
+++ b/src/common/vulkan/vulkan_icd.cpp
@@ -15,8 +15,13 @@
 #include "common/Optional.h"
 #include "common/bitset_utils.h"
 #include "common/debug.h"
+#include "common/platform.h"
 #include "common/system_utils.h"
 
+#if defined(ANGLE_PLATFORM_LINUX)
+#    include <sys/sysmacros.h>
+#endif
+
 namespace
 {
 void ResetEnvironmentVar(const char *variableName, const Optional<std::string> &value)
@@ -104,6 +109,24 @@ ICDFilterFunc GetFilterForICD(vk::ICD preferredICD)
     }
 }
 
+void GetPhysicalDeviceProperties2Wrapper(
+    PFN_vkGetPhysicalDeviceProperties pGetPhysicalDeviceProperties,
+    PFN_vkGetPhysicalDeviceProperties2KHR pGetPhysicalDeviceProperties2KHR,
+    VkPhysicalDevice physicalDevice,
+    VkPhysicalDeviceProperties2KHR *physicalDeviceProperties2Out)
+{
+    if (pGetPhysicalDeviceProperties2KHR)
+    {
+        pGetPhysicalDeviceProperties2KHR(physicalDevice, physicalDeviceProperties2Out);
+    }
+    else
+    {
+        ASSERT(pGetPhysicalDeviceProperties);
+        ASSERT(physicalDeviceProperties2Out->pNext == nullptr);
+        pGetPhysicalDeviceProperties(physicalDevice, &physicalDeviceProperties2Out->properties);
+    }
+}
+
 }  // namespace
 
 // If we're loading the vulkan layers, we could be running from any random directory.
@@ -221,23 +244,68 @@ bool ScopedVkLoaderEnvironment::setICDEnvironment(const char *icd)
     return mChangedICDEnv;
 }
 
-void ChoosePhysicalDevice(PFN_vkGetPhysicalDeviceProperties pGetPhysicalDeviceProperties,
-                          const std::vector<VkPhysicalDevice> &physicalDevices,
-                          vk::ICD preferredICD,
-                          uint32_t preferredVendorID,
-                          uint32_t preferredDeviceID,
-                          VkPhysicalDevice *physicalDeviceOut,
-                          VkPhysicalDeviceProperties *physicalDevicePropertiesOut)
+#define vkGetPhysicalDeviceProperties2KHR(physicalDevice, physicalDevicePropertiesOut)      \
+    GetPhysicalDeviceProperties2Wrapper(pGetPhysicalDeviceProperties,                       \
+                                        pGetPhysicalDeviceProperties2KHR, (physicalDevice), \
+                                        (physicalDevicePropertiesOut))
+
+void ChoosePhysicalDevice(
+    PFN_vkGetPhysicalDeviceProperties pGetPhysicalDeviceProperties,
+    PFN_vkGetPhysicalDeviceProperties2KHR pGetPhysicalDeviceProperties2KHR,
+    PFN_vkEnumerateDeviceExtensionProperties pEnumerateDeviceExtensionProperties,
+    const std::vector<VkPhysicalDevice> &physicalDevices,
+    vk::ICD preferredICD,
+    uint64_t systemDeviceID,
+    VkPhysicalDevice *physicalDeviceOut,
+    VkPhysicalDeviceProperties *physicalDevicePropertiesOut)
 {
     ASSERT(!physicalDevices.empty());
 
     ICDFilterFunc filter = GetFilterForICD(preferredICD);
 
-    const bool shouldChooseByID = (preferredVendorID != 0 || preferredDeviceID != 0);
+    const bool shouldChooseByID = (systemDeviceID != 0);
 
     for (const VkPhysicalDevice &physicalDevice : physicalDevices)
     {
-        pGetPhysicalDeviceProperties(physicalDevice, physicalDevicePropertiesOut);
+        VkPhysicalDeviceProperties2KHR physicalDeviceProperties2;
+        physicalDeviceProperties2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2_KHR;
+        physicalDeviceProperties2.pNext = nullptr;
+
+#if defined(ANGLE_PLATFORM_LINUX)
+        VkPhysicalDeviceDrmPropertiesEXT drmProperties;
+        if (shouldChooseByID)
+        {
+            uint32_t extensionCount;
+            VkResult result = pEnumerateDeviceExtensionProperties(physicalDevice, nullptr,
+                                                                  &extensionCount, nullptr);
+            ASSERT(result == VK_SUCCESS);
+            std::vector<VkExtensionProperties> extensionProperties(extensionCount);
+            result = pEnumerateDeviceExtensionProperties(physicalDevice, nullptr, &extensionCount,
+                                                         extensionProperties.data());
+            ASSERT(result == VK_SUCCESS);
+            bool isPhysicalDeviceDrmAvailable = false;
+            for (const auto &extension : extensionProperties)
+            {
+                if (strcmp(extension.extensionName, VK_EXT_PHYSICAL_DEVICE_DRM_EXTENSION_NAME) == 0)
+                {
+                    isPhysicalDeviceDrmAvailable = true;
+                    break;
+                }
+            }
+            if (isPhysicalDeviceDrmAvailable && pGetPhysicalDeviceProperties2KHR)
+            {
+                drmProperties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DRM_PROPERTIES_EXT;
+                drmProperties.pNext = nullptr;
+                physicalDeviceProperties2.pNext = &drmProperties;
+            }
+            else
+            {
+                continue;
+            }
+        }
+#endif
+        vkGetPhysicalDeviceProperties2KHR(physicalDevice, &physicalDeviceProperties2);
+        *physicalDevicePropertiesOut = physicalDeviceProperties2.properties;
 
         if (physicalDevicePropertiesOut->apiVersion < kMinimumVulkanAPIVersion)
         {
@@ -254,28 +322,16 @@ void ChoosePhysicalDevice(PFN_vkGetPhysicalDeviceProperties pGetPhysicalDevicePr
 
         if (shouldChooseByID)
         {
-            // NOTE: If the system has multiple GPUs with the same vendor and
-            // device IDs, this will arbitrarily select one of them.
-            bool matchVendorID = true;
-            bool matchDeviceID = true;
-
-            if (preferredVendorID != 0 &&
-                preferredVendorID != physicalDevicePropertiesOut->vendorID)
-            {
-                matchVendorID = false;
-            }
-
-            if (preferredDeviceID != 0 &&
-                preferredDeviceID != physicalDevicePropertiesOut->deviceID)
-            {
-                matchDeviceID = false;
-            }
-
-            if (matchVendorID && matchDeviceID)
+#if defined(ANGLE_PLATFORM_LINUX)
+            dev_t drmRenderNodeDeviceId = static_cast<dev_t>(systemDeviceID);
+            if (drmProperties.hasRender == VK_TRUE &&
+                drmProperties.renderMajor == major(drmRenderNodeDeviceId) &&
+                drmProperties.renderMinor == minor(drmRenderNodeDeviceId))
             {
                 *physicalDeviceOut = physicalDevice;
                 return;
             }
+#endif
         }
     }
 
@@ -283,7 +339,11 @@ void ChoosePhysicalDevice(PFN_vkGetPhysicalDeviceProperties pGetPhysicalDevicePr
     VkPhysicalDeviceProperties integratedDeviceProperties;
     for (const VkPhysicalDevice &physicalDevice : physicalDevices)
     {
-        pGetPhysicalDeviceProperties(physicalDevice, physicalDevicePropertiesOut);
+        VkPhysicalDeviceProperties2KHR physicalDeviceProperties2;
+        physicalDeviceProperties2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2_KHR;
+        physicalDeviceProperties2.pNext = nullptr;
+        vkGetPhysicalDeviceProperties2KHR(physicalDevice, &physicalDeviceProperties2);
+        *physicalDevicePropertiesOut = physicalDeviceProperties2.properties;
 
         if (physicalDevicePropertiesOut->apiVersion < kMinimumVulkanAPIVersion)
         {
@@ -318,9 +378,15 @@ void ChoosePhysicalDevice(PFN_vkGetPhysicalDeviceProperties pGetPhysicalDevicePr
     WARN() << "Preferred device ICD not found. Using default physicalDevice instead.";
     // Fallback to the first device.
     *physicalDeviceOut = physicalDevices[0];
-    pGetPhysicalDeviceProperties(*physicalDeviceOut, physicalDevicePropertiesOut);
+    VkPhysicalDeviceProperties2KHR physicalDeviceProperties2;
+    physicalDeviceProperties2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2_KHR;
+    physicalDeviceProperties2.pNext = nullptr;
+    vkGetPhysicalDeviceProperties2KHR(*physicalDeviceOut, &physicalDeviceProperties2);
+    *physicalDevicePropertiesOut = physicalDeviceProperties2.properties;
 }
 
+#undef vkGetPhysicalDeviceProperties2KHR
+
 }  // namespace vk
 
 }  // namespace angle
diff --git a/src/common/vulkan/vulkan_icd.h b/src/common/vulkan/vulkan_icd.h
index d887e15efa82acde28823a4609eff2f5f29e8157..715ee17a8d64a6d8acadfa353efc2c5b093b28b0 100644
--- a/src/common/vulkan/vulkan_icd.h
+++ b/src/common/vulkan/vulkan_icd.h
@@ -61,13 +61,15 @@ class [[nodiscard]] ScopedVkLoaderEnvironment : angle::NonCopyable
     Optional<std::string> mPreviousNoDeviceSelectEnv;
 };
 
-void ChoosePhysicalDevice(PFN_vkGetPhysicalDeviceProperties pGetPhysicalDeviceProperties,
-                          const std::vector<VkPhysicalDevice> &physicalDevices,
-                          vk::ICD preferredICD,
-                          uint32_t preferredVendorID,
-                          uint32_t preferredDeviceID,
-                          VkPhysicalDevice *physicalDeviceOut,
-                          VkPhysicalDeviceProperties *physicalDevicePropertiesOut);
+void ChoosePhysicalDevice(
+    PFN_vkGetPhysicalDeviceProperties pGetPhysicalDeviceProperties,
+    PFN_vkGetPhysicalDeviceProperties2KHR pGetPhysicalDeviceProperties2KHR,
+    PFN_vkEnumerateDeviceExtensionProperties pEnumerateDeviceExtensionProperties,
+    const std::vector<VkPhysicalDevice> &physicalDevices,
+    vk::ICD preferredICD,
+    uint64_t systemDeviceID,
+    VkPhysicalDevice *physicalDeviceOut,
+    VkPhysicalDeviceProperties *physicalDevicePropertiesOut);
 
 }  // namespace vk
 
diff --git a/src/libANGLE/Display.cpp b/src/libANGLE/Display.cpp
index 745a7e7c99c831b50963e785b04323070039e36e..763295b4e3e16852707931fb58682f73042916d7 100644
--- a/src/libANGLE/Display.cpp
+++ b/src/libANGLE/Display.cpp
@@ -2232,8 +2232,7 @@ static ClientExtensions GenerateClientExtensions()
 #endif
 
 #if defined(ANGLE_ENABLE_VULKAN)
-    extensions.platformANGLEVulkan   = true;
-    extensions.platformANGLEDeviceId = true;
+    extensions.platformANGLEVulkan = true;
 #endif
 
 #if defined(ANGLE_ENABLE_SWIFTSHADER)
@@ -2261,6 +2260,10 @@ static ClientExtensions GenerateClientExtensions()
     extensions.displayPowerPreferenceANGLE = true;
 #endif
 
+#if defined(ANGLE_PLATFORM_LINUX) && (defined(ANGLE_ENABLE_OPENGL) || defined(ANGLE_ENABLE_VULKAN))
+    extensions.platformANGLEDeviceId = true;
+#endif
+
     extensions.clientGetAllProcAddresses = true;
     extensions.debug                     = true;
     extensions.featureControlANGLE       = true;
diff --git a/src/libANGLE/renderer/gl/egl/DisplayEGL.cpp b/src/libANGLE/renderer/gl/egl/DisplayEGL.cpp
index d8a986e687588e91c6e6475efecc39855951c9a9..73814866877196e8248faa818007384b67de1e47 100644
--- a/src/libANGLE/renderer/gl/egl/DisplayEGL.cpp
+++ b/src/libANGLE/renderer/gl/egl/DisplayEGL.cpp
@@ -321,8 +321,8 @@ egl::Error DisplayEGL::initialize(egl::Display *display)
         reinterpret_cast<void *>(mDisplayAttributes.get(EGL_PLATFORM_ANGLE_EGL_HANDLE_ANGLE, 0));
     EGLAttrib platformType =
         mDisplayAttributes.get(EGL_PLATFORM_ANGLE_NATIVE_PLATFORM_TYPE_ANGLE, 0);
-    ANGLE_TRY(
-        mEGL->initialize(platformType, display->getNativeDisplayId(), getEGLPath(), eglHandle));
+    ANGLE_TRY(mEGL->initialize(platformType, display->getNativeDisplayId(), mDisplayAttributes,
+                               getEGLPath(), eglHandle));
 
     const gl::Version eglVersion(static_cast<uint8_t>(mEGL->majorVersion),
                                  static_cast<uint8_t>(mEGL->minorVersion));
diff --git a/src/libANGLE/renderer/gl/egl/FunctionsEGL.cpp b/src/libANGLE/renderer/gl/egl/FunctionsEGL.cpp
index 298af40088508beded76ae0fa88e83e3d2d653f1..e96ebc1d76d11f7c75c295b2fcc5e8b243b447c5 100644
--- a/src/libANGLE/renderer/gl/egl/FunctionsEGL.cpp
+++ b/src/libANGLE/renderer/gl/egl/FunctionsEGL.cpp
@@ -8,10 +8,12 @@
 
 #include "libANGLE/renderer/gl/egl/FunctionsEGL.h"
 
+#include <sys/stat.h>
 #include <algorithm>
 
 #include "common/platform.h"
 #include "common/string_utils.h"
+#include "gpu_info_util/SystemInfo.h"
 #include "libANGLE/renderer/driver_utils.h"
 #include "libANGLE/renderer/gl/FunctionsGL.h"
 #include "libANGLE/renderer/gl/egl/functionsegl_typedefs.h"
@@ -195,7 +197,9 @@ FunctionsEGL::~FunctionsEGL()
     SafeDelete(mFnPtrs);
 }
 
-egl::Error FunctionsEGL::initialize(EGLAttrib platformType, EGLNativeDisplayType nativeDisplay)
+egl::Error FunctionsEGL::initialize(EGLAttrib platformType,
+                                    EGLNativeDisplayType nativeDisplay,
+                                    const egl::AttributeMap &displayAttributes)
 {
 #define ANGLE_GET_PROC_OR_WARNING(MEMBER, NAME)                \
     do                                                         \
@@ -238,20 +242,38 @@ egl::Error FunctionsEGL::initialize(EGLAttrib platformType, EGLNativeDisplayType
     ANGLE_GET_PROC_OR_ERROR(&mFnPtrs->surfaceAttribPtr, eglSurfaceAttrib);
     ANGLE_GET_PROC_OR_ERROR(&mFnPtrs->swapIntervalPtr, eglSwapInterval);
 
-    if (IsValidPlatformTypeForPlatformDisplayConnection(platformType))
+    uint32_t systemDeviceIDHigh =
+        static_cast<uint32_t>(displayAttributes.get(EGL_PLATFORM_ANGLE_DEVICE_ID_HIGH_ANGLE, 0));
+    uint32_t systemDeviceIDLow =
+        static_cast<uint32_t>(displayAttributes.get(EGL_PLATFORM_ANGLE_DEVICE_ID_LOW_ANGLE, 0));
+    uint64_t systemDeviceID =
+        angle::GetSystemDeviceIdFromParts(systemDeviceIDHigh, systemDeviceIDLow);
+
+    if (platformType == EGL_PLATFORM_SURFACELESS_MESA && systemDeviceID != 0)
     {
-        mEGLDisplay = getPlatformDisplay(platformType, nativeDisplay);
-    }
-    else
-    {
-        mEGLDisplay = mFnPtrs->getDisplayPtr(nativeDisplay);
+        // If a native device ID was provided for surfaceless, try looking for that specific device
+        // first.
+        mEGLDisplay = getNativeDisplay(&minorVersion, &majorVersion, systemDeviceID);
     }
 
-    if (mEGLDisplay != EGL_NO_DISPLAY &&
-        mFnPtrs->initializePtr(mEGLDisplay, &majorVersion, &minorVersion) != EGL_TRUE)
+    if (mEGLDisplay == EGL_NO_DISPLAY)
     {
-        mEGLDisplay = EGL_NO_DISPLAY;
+        if (IsValidPlatformTypeForPlatformDisplayConnection(platformType))
+        {
+            mEGLDisplay = getPlatformDisplay(platformType, nativeDisplay);
+        }
+        else
+        {
+            mEGLDisplay = mFnPtrs->getDisplayPtr(nativeDisplay);
+        }
+
+        if (mEGLDisplay != EGL_NO_DISPLAY &&
+            mFnPtrs->initializePtr(mEGLDisplay, &majorVersion, &minorVersion) != EGL_TRUE)
+        {
+            mEGLDisplay = EGL_NO_DISPLAY;
+        }
     }
+
     if (mEGLDisplay == EGL_NO_DISPLAY)
     {
         // If no display was available, try to fallback to the first available
@@ -432,7 +454,7 @@ EGLDisplay FunctionsEGL::getPlatformDisplay(EGLAttrib platformType,
                                     reinterpret_cast<void *>(nativeDisplay), nullptr);
 }
 
-EGLDisplay FunctionsEGL::getNativeDisplay(int *major, int *minor)
+EGLDisplay FunctionsEGL::getNativeDisplay(int *major, int *minor, uint64_t systemDeviceID)
 {
     // We haven't queried extensions yet since some platforms require a display
     // to do so. We'll query them now since we need some for this fallback, and
@@ -444,10 +466,11 @@ EGLDisplay FunctionsEGL::getNativeDisplay(int *major, int *minor)
     }
     angle::SplitStringAlongWhitespace(extensions, &mExtensions);
 
-    // This fallback mechanism makes use of:
+    // This function makes use of:
     // - EGL_EXT_device_enumeration or EGL_EXT_device_base for eglQueryDevicesEXT
     // - EGL_EXT_platform_base for eglGetPlatformDisplayEXT
     // - EGL_EXT_platform_device for EGL_PLATFORM_DEVICE_EXT
+    // - EGL_EXT_device_query for eglQueryDeviceStringEXT if systemDeviceID is non-zero
     PFNEGLQUERYDEVICESEXTPROC queryDevices;
     PFNEGLGETPLATFORMDISPLAYEXTPROC getPlatformDisplay;
     bool hasQueryDevicesEXT =
@@ -464,6 +487,20 @@ EGLDisplay FunctionsEGL::getNativeDisplay(int *major, int *minor)
         return EGL_NO_DISPLAY;
     }
 
+    PFNEGLQUERYDEVICESTRINGEXTPROC queryDeviceString = nullptr;
+    if (systemDeviceID != 0)
+    {
+        bool hasQueryDeviceStringEXT = hasExtension("EGL_EXT_device_query");
+        if (!hasQueryDeviceStringEXT)
+        {
+            return EGL_NO_DISPLAY;
+        }
+        if (!SetPtr(&queryDeviceString, getProcAddress("eglQueryDeviceStringEXT")))
+        {
+            return EGL_NO_DISPLAY;
+        }
+    }
+
     // Get a list of native device objects.
     const EGLint kMaxDevices = 32;
     EGLDeviceEXT eglDevices[kMaxDevices];
@@ -476,6 +513,24 @@ EGLDisplay FunctionsEGL::getNativeDisplay(int *major, int *minor)
     // Look for the first native device that gives us a valid display.
     for (EGLint i = 0; i < numDevices; i++)
     {
+        if (systemDeviceID != 0)
+        {
+#if defined(ANGLE_PLATFORM_LINUX)
+            // We are looking for a specific DRM device.
+            const char *renderNodePath =
+                queryDeviceString(eglDevices[i], EGL_DRM_RENDER_NODE_FILE_EXT);
+            if (!renderNodePath)
+                continue;
+            struct stat sb;
+            if (stat(renderNodePath, &sb) < 0)
+                continue;
+            if (sb.st_rdev != systemDeviceID)
+                continue;
+#else
+            // The other EGL using platform (Android) does not have a concept of a system device ID.
+            continue;
+#endif
+        }
         EGLDisplay display = getPlatformDisplay(EGL_PLATFORM_DEVICE_EXT, eglDevices[i], nullptr);
         if (mFnPtrs->getErrorPtr() != EGL_SUCCESS)
         {
diff --git a/src/libANGLE/renderer/gl/egl/FunctionsEGL.h b/src/libANGLE/renderer/gl/egl/FunctionsEGL.h
index 44c060389750dc672a6d02e1b414c7fe842d0880..4161d42415c459c68b4e196e40e002a58bfca018 100644
--- a/src/libANGLE/renderer/gl/egl/FunctionsEGL.h
+++ b/src/libANGLE/renderer/gl/egl/FunctionsEGL.h
@@ -15,6 +15,7 @@
 #include <string>
 #include <vector>
 
+#include "libANGLE/AttributeMap.h"
 #include "libANGLE/Error.h"
 
 namespace rx
@@ -33,7 +34,9 @@ class FunctionsEGL
     std::string vendorString;
     std::string versionString;
 
-    egl::Error initialize(EGLAttrib platformType, EGLNativeDisplayType nativeDisplay);
+    egl::Error initialize(EGLAttrib platformType,
+                          EGLNativeDisplayType nativeDisplay,
+                          const egl::AttributeMap &displayAttributes);
     egl::Error terminate();
 
     virtual void *getProcAddress(const char *name) const = 0;
@@ -132,7 +135,7 @@ class FunctionsEGL
     EGLDisplay getPlatformDisplay(EGLAttrib platformType, EGLNativeDisplayType nativeDisplay);
 
     // Fallback mechanism for creating a display from a native device object.
-    EGLDisplay getNativeDisplay(int *major, int *minor);
+    EGLDisplay getNativeDisplay(int *major, int *minor, uint64_t systemDeviceID = 0);
 
     struct EGLDispatchTable;
     EGLDispatchTable *mFnPtrs;
diff --git a/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.cpp b/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.cpp
index ddf845523a280daaaf029fe4838c703c398894b3..12f2335ff0e4e42ee82384547f02adbfd3f7eaad 100644
--- a/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.cpp
+++ b/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.cpp
@@ -30,6 +30,7 @@ FunctionsEGLDL::~FunctionsEGLDL() {}
 
 egl::Error FunctionsEGLDL::initialize(EGLAttrib platformType,
                                       EGLNativeDisplayType nativeDisplay,
+                                      const egl::AttributeMap &displayAttributes,
                                       const char *libName,
                                       void *eglHandle)
 {
@@ -59,7 +60,7 @@ egl::Error FunctionsEGLDL::initialize(EGLAttrib platformType,
         return egl::Error(EGL_NOT_INITIALIZED, "Could not find eglGetProcAddress");
     }
 
-    return FunctionsEGL::initialize(platformType, nativeDisplay);
+    return FunctionsEGL::initialize(platformType, nativeDisplay, displayAttributes);
 }
 
 void *FunctionsEGLDL::getProcAddress(const char *name) const
diff --git a/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.h b/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.h
index 22cfaf952404fd7d0dd1956708933974328857a0..dda6522de0a74da20dbe57481cd4d0095cbb1faf 100644
--- a/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.h
+++ b/src/libANGLE/renderer/gl/egl/FunctionsEGLDL.h
@@ -22,6 +22,7 @@ class FunctionsEGLDL : public FunctionsEGL
 
     egl::Error initialize(EGLAttrib platformType,
                           EGLNativeDisplayType nativeDisplay,
+                          const egl::AttributeMap &displayAttributes,
                           const char *libName,
                           void *eglHandle);
     void *getProcAddress(const char *name) const override;
diff --git a/src/libANGLE/renderer/vulkan/DisplayVk.cpp b/src/libANGLE/renderer/vulkan/DisplayVk.cpp
index 70d53d091f8baddb329ccdafac4cbe6847549ead..28f8ba20a02430ddae1336746f31e8bd3c6f8712 100644
--- a/src/libANGLE/renderer/vulkan/DisplayVk.cpp
+++ b/src/libANGLE/renderer/vulkan/DisplayVk.cpp
@@ -11,6 +11,7 @@
 
 #include "common/debug.h"
 #include "common/system_utils.h"
+#include "gpu_info_util/SystemInfo.h"
 #include "libANGLE/BlobCache.h"
 #include "libANGLE/Context.h"
 #include "libANGLE/Display.h"
@@ -170,14 +171,16 @@ egl::Error DisplayVk::initialize(egl::Display *display)
 
     const vk::UseDebugLayers useDebugLayers = ShouldLoadDebugLayers(attribs);
     const angle::vk::ICD desiredICD         = ChooseICDFromAttribs(attribs);
-    const uint32_t preferredVendorId =
+    const uint32_t systemDeviceIdHigh =
         static_cast<uint32_t>(attribs.get(EGL_PLATFORM_ANGLE_DEVICE_ID_HIGH_ANGLE, 0));
-    const uint32_t preferredDeviceId =
+    const uint32_t systemDeviceIdLow =
         static_cast<uint32_t>(attribs.get(EGL_PLATFORM_ANGLE_DEVICE_ID_LOW_ANGLE, 0));
+    const uint64_t systemDeviceId =
+        angle::GetSystemDeviceIdFromParts(systemDeviceIdHigh, systemDeviceIdLow);
 
     angle::Result result = mRenderer->initialize(
-        this, this, desiredICD, preferredVendorId, preferredDeviceId, useDebugLayers,
-        getWSIExtension(), getWSILayer(), getWindowSystem(), mState.featureOverrides);
+        this, this, desiredICD, systemDeviceId, useDebugLayers, getWSIExtension(),
+        getWSILayer(), getWindowSystem(), mState.featureOverrides);
     ANGLE_TRY(angle::ToEGL(result, EGL_NOT_INITIALIZED));
 
     mDeviceQueueIndex = mRenderer->getDeviceQueueIndex(egl::ContextPriority::Medium);
diff --git a/src/libANGLE/renderer/vulkan/vk_renderer.cpp b/src/libANGLE/renderer/vulkan/vk_renderer.cpp
index 6579b574316dbb545ba8ee3fb79d3094005364f2..3ac632e25fc3ab3486411e77aceb7d78333ee5b3 100644
--- a/src/libANGLE/renderer/vulkan/vk_renderer.cpp
+++ b/src/libANGLE/renderer/vulkan/vk_renderer.cpp
@@ -2457,8 +2457,7 @@ angle::Result Renderer::enableInstanceExtensions(vk::ErrorContext *context,
 angle::Result Renderer::initialize(vk::ErrorContext *context,
                                    vk::GlobalOps *globalOps,
                                    angle::vk::ICD desiredICD,
-                                   uint32_t preferredVendorId,
-                                   uint32_t preferredDeviceId,
+                                   uint64_t systemDeviceId,
                                    UseDebugLayers useDebugLayers,
                                    const char *wsiExtension,
                                    const char *wsiLayer,
@@ -2686,9 +2685,9 @@ angle::Result Renderer::initialize(vk::ErrorContext *context,
     std::vector<VkPhysicalDevice> physicalDevices(physicalDeviceCount);
     ANGLE_VK_TRY(context, vkEnumeratePhysicalDevices(mInstance, &physicalDeviceCount,
                                                      physicalDevices.data()));
-    ChoosePhysicalDevice(vkGetPhysicalDeviceProperties, physicalDevices, mEnabledICD,
-                         preferredVendorId, preferredDeviceId, &mPhysicalDevice,
-                         &mPhysicalDeviceProperties);
+    ChoosePhysicalDevice(vkGetPhysicalDeviceProperties, vkGetPhysicalDeviceProperties2KHR,
+                         vkEnumerateDeviceExtensionProperties, physicalDevices, mEnabledICD,
+                         systemDeviceId, &mPhysicalDevice, &mPhysicalDeviceProperties);
 
     // The device version that is assumed by ANGLE is the minimum of the actual device version and
     // the highest it's allowed to use.
diff --git a/src/libANGLE/renderer/vulkan/vk_renderer.h b/src/libANGLE/renderer/vulkan/vk_renderer.h
index 765ed40f4e68ab091c42d85b6e9af19be3048ffb..3a8b6f1669fb1f5832ba3e0dde88857117c85b27 100644
--- a/src/libANGLE/renderer/vulkan/vk_renderer.h
+++ b/src/libANGLE/renderer/vulkan/vk_renderer.h
@@ -144,8 +144,7 @@ class Renderer : angle::NonCopyable
     angle::Result initialize(vk::ErrorContext *context,
                              vk::GlobalOps *globalOps,
                              angle::vk::ICD desiredICD,
-                             uint32_t preferredVendorId,
-                             uint32_t preferredDeviceId,
+                             uint64_t systemDeviceId,
                              UseDebugLayers useDebugLayers,
                              const char *wsiExtension,
                              const char *wsiLayer,
diff --git a/src/tests/test_utils/VulkanHelper.cpp b/src/tests/test_utils/VulkanHelper.cpp
index c5cd469ca4d1fa79def3a75e86c847b1593c3826..257f15291bac20239634d22d0308548e922b343a 100644
--- a/src/tests/test_utils/VulkanHelper.cpp
+++ b/src/tests/test_utils/VulkanHelper.cpp
@@ -264,7 +264,8 @@ void VulkanHelper::initialize(bool useSwiftshader, bool enableValidationLayers)
     ASSERT(physicalDevices.size() > 0);
 
     VkPhysicalDeviceProperties physicalDeviceProperties;
-    ChoosePhysicalDevice(vkGetPhysicalDeviceProperties, physicalDevices, icd, 0, 0,
+    ChoosePhysicalDevice(vkGetPhysicalDeviceProperties, vkGetPhysicalDeviceProperties2,
+                         vkEnumerateDeviceExtensionProperties, physicalDevices, icd, 0,
                          &mPhysicalDevice, &physicalDeviceProperties);
 
     vkGetPhysicalDeviceMemoryProperties(mPhysicalDevice, &mMemoryProperties);
@@ -461,6 +462,10 @@ void VulkanHelper::initializeFromANGLE()
     vkResult = vkCreateCommandPool(mDevice, &commandPoolCreateInfo, nullptr, &mCommandPool);
     ASSERT(vkResult == VK_SUCCESS);
 
+    vkGetPhysicalDeviceProperties2 = reinterpret_cast<PFN_vkGetPhysicalDeviceProperties2>(
+        vkGetInstanceProcAddr(mInstance, "vkGetPhysicalDeviceProperties2"));
+    ASSERT(vkGetPhysicalDeviceProperties2);
+
     vkGetPhysicalDeviceImageFormatProperties2 =
         reinterpret_cast<PFN_vkGetPhysicalDeviceImageFormatProperties2>(
             vkGetInstanceProcAddr(mInstance, "vkGetPhysicalDeviceImageFormatProperties2"));
diff --git a/src/tests/test_utils/VulkanHelper.h b/src/tests/test_utils/VulkanHelper.h
index bf288c75bebe53ef58b41954f139be74bbf80b7e..3958b1f0a21553ec76074aa1890ee36474722421 100644
--- a/src/tests/test_utils/VulkanHelper.h
+++ b/src/tests/test_utils/VulkanHelper.h
@@ -177,6 +177,7 @@ class VulkanHelper
     bool mHasExternalSemaphoreFd      = false;
     bool mHasExternalSemaphoreFuchsia = false;
     bool mHasUnifiedImageLayouts      = false;
+    PFN_vkGetPhysicalDeviceProperties2 vkGetPhysicalDeviceProperties2 = nullptr;
     PFN_vkGetPhysicalDeviceImageFormatProperties2 vkGetPhysicalDeviceImageFormatProperties2 =
         nullptr;
     PFN_vkGetMemoryFdKHR vkGetMemoryFdKHR       = nullptr;
